zcutil/fetch-params.sh

```diff
@@ -6,27 +6,61 @@ PARAMS_DIR="$HOME/.zcash-params"
 
 SPROUT_PKEY_NAME='sprout-proving.key'
 SPROUT_VKEY_NAME='sprout-verifying.key'
-SPROUT_PKEY_URL="https://z.cash/downloads/$SPROUT_PKEY_NAME"
-SPROUT_VKEY_URL="https://z.cash/downloads/$SPROUT_VKEY_NAME"
+SPROUT_URL="https://z.cash/downloads"
+SPROUT_IPFS="/ipfs/QmZKKx7Xup7LiAtFRhYsE1M7waXcv9ir9eCECyXAFGxhEo"
 
 SHA256CMD="$(command -v sha256sum || echo shasum)"
 SHA256ARGS="$(command -v sha256sum >/dev/null || echo '-a 256')"
 
+WGETCMD="$(command -v wget)"
+IPFSCMD="$(command -v ipfs)"
+
+# fetch methods can be disabled with ZC_DISABLE_SOMETHING=1
+ZC_DISABLE_WGET="${ZC_DISABLE_WGET:-}"
+ZC_DISABLE_IPFS="${ZC_DISABLE_IPFS:-}"
+
+function fetch_wget {
+    if [ -z "$WGETCMD" ] || ! [ -z "$ZC_DISABLE_WGET" ]; then
+        return 1
+    fi
+
+    local filename="$1"
+    local dlname="$2"
+
+    wget \
+        --progress=dot:giga \
+        --output-document="$dlname" \
+        --continue \
+        --retry-connrefused --waitretry=3 --timeout=30 \
+        "$SPROUT_URL/$filename"
+}
+
+function fetch_ipfs {
+    if [ -z "$IPFSCMD" ] || ! [ -z "$ZC_DISABLE_IPFS" ]; then
+        return 1
+    fi
+
+    local filename="$1"
+    local dlname="$2"
+
+    ipfs get --output "$dlname" "$SPROUT_IPFS/$filename"
+}
+
 function fetch_params {
-    local url="$1"
+    local filename="$1"
     local output="$2"
     local dlname="${output}.dl"
     local expectedhash="$3"
 
     if ! [ -f "$output" ]
     then
-        echo "Retrieving: $url"
-        wget \
-            --progress=dot:giga \
-            --output-document="$dlname" \
-            --continue \
-            --retry-connrefused --waitretry=3 --timeout=30 \
-            "$url"
+        for method in ipfs wget; do
+            echo "Retrieving($method): $filename"
```

Comment:
Missing a space?

Also, I think this should move into the fetch functions, just after the "Should I try this" check, so that if someone doesn't have IPFS installed then it doesn't show up. Or, there needs to be clearer indication when a method fails.

Reply:
The missing space in `Retrieving($method)` was intentional. I've added it and moved the echo to the fetch_* functions, making the output more specific as well.